clients/go/lanyard/client.go

```diff
@@ -0,0 +1,255 @@
+// An API client for [lanyard.org].
+//
+// [lanyard.org]: https://lanyard.org
+package lanyard
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/contextwtf/lanyard/types"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"golang.org/x/xerrors"
+)
+
+var ErrNotFound error = xerrors.New("resource not found")
+
+type Client struct {
+	httpClient *http.Client
+	url        string
+}
+
+type ClientOpt func(*Client)
+
+func WithURL(url string) ClientOpt {
+	return func(c *Client) {
+		c.url = url
+	}
+}
+
+func WithClient(hc *http.Client) ClientOpt {
+	return func(c *Client) {
+		c.httpClient = hc
+	}
+}
+
+// Uses https://lanyard.org/api/v1 for a default url
+// and http.Client with a 30s timeout unless specified
+// using [WithURL] or [WithClient]
+func New(opts ...ClientOpt) *Client {
+	const url = "https://lanyard.org/api/v1"
+	c := &Client{
+		url: url,
+		httpClient: &http.Client{
+			Timeout: 30 * time.Second,
+		},
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	return c
+}
+
+func (c *Client) sendRequest(
+	ctx context.Context,
+	method, path string,
+	body, destination any,
+) error {
+	var (
+		req *http.Request
+		err error
+	)
+
+	url := c.url + path
+
+	if body == nil {
+		req, err = http.NewRequestWithContext(ctx, method, url, nil)
+
+		if err != nil {
+			return xerrors.Errorf("error creating request: %w", err)
+		}
+	} else {
+		b, err := json.Marshal(body)
+		if err != nil {
+			return xerrors.Errorf("failed to marshal body: %w", err)
+		}
+		req, err = http.NewRequestWithContext(ctx, method, url, bytes.NewReader(b))
+		if err != nil {
+			return xerrors.Errorf("error creating request: %w", err)
+		}
+
+		req.Header.Set("Content-Type", "application/json")
+	}
+
+	req.Header.Set("User-Agent", "lanyard-go+v1.0.0")
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return xerrors.Errorf("failed to send request: %w", err)
+	}
+
+	if resp.StatusCode >= 400 {
+		// special case 404s to make consuming client API easier
+		if resp.StatusCode == http.StatusNotFound {
+			return ErrNotFound
+		}
+
+		return xerrors.Errorf("error making http request: %s", resp.Status)
+	}
+
+	defer resp.Body.Close()
+
+	if err := json.NewDecoder(resp.Body).Decode(&destination); err != nil {
+		return xerrors.Errorf("failed to decode response: %w", err)
+	}
+
+	return nil
+}
+
+type CreateTreeRequest struct {
+	// UnhashedLeaves is a slice of addresses or ABI encoded types
+	UnhashedLeaves []hexutil.Bytes `json:"unhashedLeaves"`
+
+	// LeafTypeDescriptor describes the abi-encoded types of the leaves, and
+	// is required if leaves are not address types
+	LeafTypeDescriptor []string `json:"leafTypeDescriptor,omitempty"`
+
+	// PackedEncoding is true by default
+	PackedEncoding types.JsonNullBool `json:"packedEncoding,omitempty"` // what's sent over the wire
```

Comment:
What behavior are we looking for with PackedEncoding?

Reply:
ah, that comment was old, removed!

Reply:
(otherwise just want to maintain same behavior as the http api0